andela/ah-backend-poseidon
# #162163001 Enable Social login via Google, Facebook and Twitter

# **What does this PR do?**
- Adds social login for google, facebook and twitter using [django-rest-framework-social-oauth2](https://github.com/RealmTeam/django-rest-framework-social-oauth2)
- Modify settings to add social oauth settings. 
- Modify login to use username instead of email. 
- Modify token to include username and decode username to ensure token is valid.
- Change all db to accept blank email addresses from social oauth.

**Description of the task to be completed?**

- Currently a new user cannot login with their pre-existing social media accounts. He/She can only register and or login with email password combinations. Please not that implementation in this case means STRICTLY BACKEND implementation only!

**How should this be manually tested?**
- Pull the branch to your remote workspace.
- Run migrations with the command `python manage.py migrate`. 
- Create a superuser by running the command `python manage.py createsuperuser` and follow the prompts. 
- Run the server using `python manage.py runserver`
- Open a browser and go the url 127.0.0.1:8000/admin. 
- Login with the details of the superuser you created earlier. 
- Under django oauth kit create an application with the following settings. 
```
- client_id and client_secret shouldn't be changed
- user should be your superuser
- redirect_uris should be left blank
- client_type should be set to confidential
- authorization_grant_type should be set to 'Resource owner password-based'
- name can be set to whatever you want
- see screenshot attached. 
```
![image](https://user-images.githubusercontent.com/17021137/49606650-fbd82780-f9a4-11e8-858b-87a920eaeeb1.png)
- Open POSTMAN.
- Add ```http://127.0.0.1:8000/api/auth/convert-token``` to the server URL.
- Add the following parameters. 
```
- grant_type and set it to convert_token
- client_id - set it to the client_id generated when you created the application in the django oauth admin. 
- client_secret - set it to the client_secret generated
- backend - set to 'facebook'
- token - add the token given to you by facebook. check developers.facebook.com
- this will add the data passed by facebook to the db. and provide you a token. 
```
![image](https://user-images.githubusercontent.com/17021137/49607325-abfa6000-f9a6-11e8-96f8-41f9c94ca8eb.png)

- Hit the route http://127.0.0.1:8000/api/user 
- Add the token provided in the previous step in the format.
 ```
Authorization: Bearer <token> 
```
![image](https://user-images.githubusercontent.com/17021137/49607410-dfd58580-f9a6-11e8-8165-91af04362865.png)


**Relevant pivotal tracker stories?**
#162163001
